In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam

## A. Build a baseline model

In [4]:
#loading data
file_path = 'concrete_data.csv'
data = pd.read_csv(file_path)

#separating features/target variable
X = data.drop(columns=['Strength'])
y = data['Strength']

#function for baseline model
def baseline_model():
    model = Sequential()
    model.add(Dense(10, input_dim=X.shape[1], activation='relu'))
    model.add(Dense(1)) 
    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

#initializing a list to store mean squared errors
mse_list = []

#repeating process 50 times
for _ in range(50):
    # Split the data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)


    model = baseline_model() #creating model


    model.fit(X_train, y_train, epochs=50, verbose=0) #training model

    #evaluating model
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)

    
    mse_list.append(mse) #append mean squared error to the list

#calculated mean and standard deviation of mean squared errors
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)

#printing mean and standard deviation of mean squared errors
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

10/10 [==============================] - 0s 2ms/step
Mean of Mean Squared Errors: 338.7213762809644
Standard Deviation of Mean Squared Errors: 241.90263932362242


## B. Normalize the data

In [6]:
#loading data
file_path = 'concrete_data.csv'
data = pd.read_csv(file_path)

#separating features/target variable
X = data.drop(columns=['Strength'])
y = data['Strength']

#normalizing features
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

#function for baseline model
def baseline_model():
    model = Sequential()
    model.add(Dense(10, input_dim=X_normalized.shape[1], activation='relu'))
    model.add(Dense(1))  # Output layer with default linear activation
    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

#initializing a list to store mean squared errors
mse_list = []

#repeating process 50 times
for _ in range(50):
    #split normalized data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.3)
    
    model = baseline_model() #creating model

    
    model.fit(X_train, y_train, epochs=50, verbose=0) #training model

     
 #evaluating model
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    

    mse_list.append(mse) #append mean squared error to the list


#calculated mean and standard deviation of mean squared errors
mean_mse_normalized = np.mean(mse_list)

#printing mean of mean squared errors
print(f"Mean of Mean Squared Errors (Normalized Data): {mean_mse_normalized}")

10/10 [==============================] - 0s 1ms/step
Mean of Mean Squared Errors (Normalized Data): 359.0386368374318


In this comparison, it appears that normalizing the data in Step B resulted in a slightly higher mean of mean squared errors compared to the non-normalized data in Step A. This difference might indicate that, in this specific scenario, normalizing the data didn't significantly improve the model's predictive performance; it might have even led to a slightly worse performance, as indicated by the marginally higher mean squared error. 

## C. Increate the number of epochs

In [7]:
#loading data
file_path = 'concrete_data.csv'
data = pd.read_csv(file_path)

#separating features/target variable
X = data.drop(columns=['Strength'])
y = data['Strength']

#normalizing features
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

#function for baseline model
def baseline_model():
    model = Sequential()
    model.add(Dense(10, input_dim=X_normalized.shape[1], activation='relu'))
    model.add(Dense(1)) 
    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

#initializing a list to store mean squared errors
mse_list = []

#repeating process 50 times
for _ in range(50):
    #split normalized data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.3)
    
    model = baseline_model() #creating model

    
    model.fit(X_train, y_train, epochs=100, verbose=0) #training model with 100 epochs
    
#evaluating model
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    

    mse_list.append(mse)  #append mean squared error to the list


#calculated mean and standard deviation of mean squared errors
mean_mse_100_epochs = np.mean(mse_list)

#printing mean of mean squared errors for 100 epochs
print(f"Mean of Mean Squared Errors (100 epochs): {mean_mse_100_epochs}")

10/10 [==============================] - 0s 3ms/step
Mean of Mean Squared Errors (100 epochs): 166.56643398852674


In Step B (50 epochs), the mean of mean squared errors was approximately 359.04.
In Step C (100 epochs), the mean of mean squared errors reduced significantly to approximately 166.57.

This reduction suggests that allowing the model to train for a greater number of epochs led to improved performance, resulting in lower error rates when predicting the concrete strength. Increasing the number of epochs provided the model with more opportunities to adjust its weights and improve its predictive ability.

## D. Increase the number of hidden layers

In [9]:
#loading data
file_path = 'concrete_data.csv'
data = pd.read_csv(file_path)

#separating features/target variable
X = data.drop(columns=['Strength'])
y = data['Strength']

#normalizing features
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

#function for baseline model with three hidden layers
def three_hidden_layers_model():
    model = Sequential()
    model.add(Dense(10, input_dim=X_normalized.shape[1], activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))  # Output layer with default linear activation
    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

#initializing a list to store mean squared errors
mse_list = []

#repeating process 50 times
for _ in range(50):
                            #split normalized data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.3)

    #creating model with three hidden layers
    model = three_hidden_layers_model()

    #training model
    model.fit(X_train, y_train, epochs=50, verbose=0)

    #evaluating model
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)

    mse_list.append(mse) #appending mean squared error to the list

#calculated mean and standard deviation of mean squared errors
mean_mse_three_layers = np.mean(mse_list)

#printing mean of mean squared errors for three hidden layers
print(f"Mean of Mean Squared Errors (Three Hidden Layers): {mean_mse_three_layers}")

10/10 [==============================] - 0s 2ms/step
Mean of Mean Squared Errors (Three Hidden Layers): 132.21949883176055


In Step B (50 epochs with a single hidden layer), the mean of mean squared errors was approximately 359.04.
In Step D (50 epochs with three hidden layers), the mean of mean squared errors reduced significantly to approximately 132.22.

The decrease in mean squared error when transitioning from a single hidden layer to three hidden layers suggests that the neural network architecture with multiple hidden layers and 50 epochs performed better in predicting concrete strength. This improvement could be due to the increased complexity and capacity of the network with additional hidden layers, allowing it to capture more intricate relationships within the data.